app/process/watcher/evm/watcher.go

```diff
@@ -17,47 +17,146 @@
+	"github.com/limechain/hedera-eth-bridge-validator/app/domain/repository"
 	"github.com/limechain/hedera-eth-bridge-validator/app/domain/service"
+	"github.com/limechain/hedera-eth-bridge-validator/app/helper/timestamp"
 	"github.com/limechain/hedera-eth-bridge-validator/app/model/transfer"
 	c "github.com/limechain/hedera-eth-bridge-validator/config"
 	"github.com/limechain/hedera-eth-bridge-validator/constants"
 	log "github.com/sirupsen/logrus"
+	"gorm.io/gorm"
 	"math/big"
 	"strconv"
 )
 
 type Watcher struct {
-	contracts service.Contracts
-	evmClient client.EVM
-	logger    *log.Entry
-	mappings  c.Assets
-	validator bool
+	repository  repository.Status
+	contracts   service.Contracts
+	evmClient   client.EVM
+	logger      *log.Entry
+	mappings    c.Assets
+	targetBlock uint64
+	validator   bool
 }
 
-func NewWatcher(contracts service.Contracts, evmClient client.EVM, mappings c.Assets, validator bool) *Watcher {
+func NewWatcher(
+	repository repository.Status,
+	contracts service.Contracts,
+	evmClient client.EVM,
+	mappings c.Assets,
+	startBlock int64,
+	validator bool) *Watcher {
+	targetBlock, err := evmClient.GetClient().BlockNumber(context.Background())
+	if err != nil {
+		log.Fatalf("Could not retrieve latest block. Error: [%s].", err)
+	}
+
+	if startBlock == 0 {
+		_, err := repository.GetLastFetchedTimestamp(contracts.Address().String())
+		if err != nil {
+			if errors.Is(err, gorm.ErrRecordNotFound) {
+				err := repository.CreateTimestamp(contracts.Address().String(), int64(targetBlock))
+				if err != nil {
+					log.Fatalf("[%s] - Failed to create Transfer Watcher timestamp. Error: [%s]", contracts.Address(), err)
+				}
+				log.Tracef("[%s] - Created new Transfer Watcher timestamp [%s]", contracts.Address(), timestamp.ToHumanReadable(int64(targetBlock)))
+			} else {
+				log.Fatalf("[%s] - Failed to fetch last Transfer Watcher timestamp. Error: [%s]", contracts.Address(), err)
+			}
+		}
+	} else {
+		err := repository.UpdateLastFetchedTimestamp(contracts.Address().String(), startBlock)
+		if err != nil {
+			log.Fatalf("[%s] - Failed to update Transfer Watcher Status timestamp. Error [%s]", contracts.Address(), err)
+		}
+		targetBlock = uint64(startBlock)
+		log.Tracef("[%s] - Updated Transfer Watcher timestamp to [%s]", contracts.Address(), timestamp.ToHumanReadable(startBlock))
+	}
 	return &Watcher{
-		contracts: contracts,
-		evmClient: evmClient,
-		logger:    c.GetLoggerFor(fmt.Sprintf("EVM Router Watcher [%s]", contracts.Address())),
-		mappings:  mappings,
-		validator: validator,
+		repository:  repository,
+		contracts:   contracts,
+		evmClient:   evmClient,
+		logger:      c.GetLoggerFor(fmt.Sprintf("EVM Router Watcher [%s]", contracts.Address())),
+		mappings:    mappings,
+		targetBlock: targetBlock,
+		validator:   validator,
 	}
 }
 
 func (ew *Watcher) Watch(queue qi.Queue) {
 	go ew.listenForEvents(queue)
+
+	ew.processPastLogs(queue)
```

Comment:
I think that we should first process past logs and then run the watcher. It think that if we let it the way it is, we might get into situation in which `listenForEvents` picks up an event and reacts to it and at the same time the `processPastLogs` queries for events after block X.
This way a TX may end up being processed twice

Reply:
The problem with the watch subscriptions is that they always start with the latest block. So, if you first process the past events and then start the watch subscriptions, you might end up with skipped events. Better process them twice than never. Also, most of the logic regarding that will revert in an error later on, which is not a real error - it will only be a duplication.